modules/kernel/src/main/scala-2.10/notebook/ReplCalculator.scala

```diff
@@ -210,6 +210,29 @@ class ReplCalculator(
         log.debug("Executing execute request")
         execute(sender(), er)
 
+      case InterruptCellRequest(killCellId) =>
+        // kill job(s) still waiting for execution to start, if any
+        val (jobsInQueueToKill, nonAffectedJobs) = queue.partition { case (_, ExecuteRequest(Some(cellIdInQueue), _, _)) =>
+          cellIdInQueue == killCellId
+        }
+        log.debug(s"Canceling $killCellId jobs still in queue (if any):\n", jobsInQueueToKill)
+        queue = nonAffectedJobs
+
+        // call cancelJobGroup even if job was in queue
+        // this is safe in recent spark versions, see SPARK-6414
+        // what if sparkContext is not available yet?
+        log.debug(s"Interrupting the cell: $killCellId")
+        val jobGroupId = JobTracking.jobGroupId(killCellId)
+        repl.evaluate(
+          s"""globalScope.sparkContext.cancelJobGroup("${jobGroupId}")""",
+          msg => sender() ! StreamResponse(msg, "stdout")
+        )
```

Comment:
there is one trade-off here. 

if cellId was found in queue, we may skip killing the jobGroup (useful in case  sparkContext is unavailable).

but if the same cell was run twice we might potentially have a job in queue, and a job with same `cell-id` running currently.